tools/git-change-exec/actions.go

```diff
@@ -0,0 +1,68 @@
+// Copyright (c) 2024 Zededa, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package main
+
+import (
+	"os"
+	"os/exec"
+	"reflect"
+	"strings"
+)
+
+type action interface {
+	match(path string) bool
+	do() error
+}
+
+func id(i any) string {
+	ty := reflect.TypeOf(i)
+	return ty.Name()
+}
+
+func execCmdWithDefaults(name string, args ...string) *exec.Cmd {
+	cmd := exec.Command(name, args...)
+
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	cmd.Stdin = os.Stdin
+
+	return cmd
+}
+
+// Do not forget to add your Action HERE
+var actions = []action{
+	pillarTestAction{},
+	gitChangeExecTest{},
+	getDepsTestAction{},
+}
+
+type pillarTestAction struct{}
+
+func (b pillarTestAction) match(path string) bool {
+	return strings.HasPrefix(path, "pkg/pillar")
+}
+
+func (b pillarTestAction) do() error {
+	return execCmdWithDefaults("make", "-C", "pkg/pillar", "test").Run()
```

Comment:
Just an idea. Instead of running commands from go, can we pass control to handlers here? And the handlers can be any script, like bash/sh ones. Like I do in memory-monitor. And we can store the handler script in the same directory.

Reply:
How would this f.e. look like? Where would the handler script be stored? In the directory of this tool or f.e. in `pkg/pillar`?

The advantage is that you can run the script to trigger an action manually? But then how would this work f.e. for https://github.com/christoph-zededa/eve/pull/1 ?

Reply:
The handlers could be in the same directory as the tool itself: `tools/git-change-exec/handlers`. Or, if it's a common tool, they can be stored wherever else. 

The advantage is that I define the action as a (bash) script, which I can easily debug, and it's just a regular way to run some "chores" workflows. I can agree that we may need some extra power of golang to define logic on how to filter the files or check conditions, but I definitely do not need golang to run actions after that. So, just define an action as a set of 1) function to filter the files / check conditions, 2) function to prepare the args for the handler and then run it with `exec.Command()`.

In the case of your PR it could be:

```go
handlerName := "tools/spdx-check.sh"
argsArgs := s.checkFiles
```

and then some generic code that runs for any action:

```go
exec.Command(handlerName, handlerArgs)
```

Reply:
But not every action needs the `handlerArgs` and perhaps it needs different arguments, like f.e. https://github.com/lf-edge/eve/pull/4202#issuecomment-2329595032 .

I fear we just create 3 shell scripts with 2 lines of license, and 1 line of executable. So when you want to add an action you do not only need to change the go code, but also have to add a shell script.

> I definitely do not need golang to run actions after that

I don't see this as a big problem. If you want to run it manually look up how the action runs.
If we would use the shell, then you would still struggle with the arguments to the shell script.

> In the case of your PR it could be:
> handlerName := "tools/spdx-check.sh"
> argsArgs := s.checkFiles

Isnt' that just syntactic sugar over writing this:
```
handlerName := "tools/spdx-check.sh"
execCmdWithDefaults(handlerName).Run()
```
(of course you can just inline `handlerName`)

---

> In the case of your PR it could be:
> handlerName := "tools/spdx-check.sh"
> argsArgs := s.checkFiles

Also I thought the script would then be in the `tools/git-change-exec/handlers`.

Reply:
> But not every action needs the handlerArgs and perhaps it needs different arguments, like f.e. https://github.com/lf-edge/eve/pull/4202#issuecomment-2329595032 .

Will it not be just a set of options for the `go test` run? Like
```go
handlerName = "go"
handlerArgs = "test -run <TestName>" // define it previously by some magic logic
```

> I fear we just create 3 shell scripts with 2 lines of license, and 1 line of executable. So when you want to add an action you do not only need to change the go code, but also have to add a shell script.

Then we should just pass the path to the executable, as in the previous example: 
```go
handlerName := "tools/spdx-check.sh"
```
, in this case, we should not require an extra wrapper.

> I don't see this as a big problem. 

I see. But the reasons are pretty subjective. I think it's just an overkill for the most of fast chore tasks. I will refrain from writing any actions in this case, as I want to avoid going through the development cycle of the write-compile-test in a language that I'm not an expert in. But maybe for others, that's fine.

Reply:
Just to be clear. I really appreciate the concept of PR; I think it's super important. Thanks for bringing it up. However, I wouldn't say I'm a big fan of the way it's presented. I know it's totally subjective, and if it works for others, that's totally fine with me too.